test/functional/testlib.bash

```diff
@@ -2659,6 +2694,429 @@ create_bundle() { # swupd_function
+	debug_msg "Looking for files added during the update..."
+
+	for filename in "${files_new_manifest[@]}"; do
+
+		debug_msg "Looking for file $filename in the previous version..."
+
+		old_hash=$(get_hash_from_manifest "$old_manifest" "$filename")
+		new_hash=$(get_hash_from_manifest "$new_manifest" "$filename")
+
+		if [ -z "$old_hash" ]; then
+			# if the file didn't exist in the old version, we need to add the
+			# whole file not just the delta
+			debug_msg "File $filename is a new file"
+			debug_msg "Adding fullfile to pack -> $new_version_dir/files/$new_hash"
+			add_to_pack "$bundle_name" "$new_version_dir"/files/"$new_hash" "$old_version"
+		fi
+
+	done
+
+}
+
+update_bundle_from_content() { #swupd_function
+
+	ubfc_usage() {
+		cat <<-EOM
+		Usage:
+		    update_bundle_from_content -n <bundle_name> -c <content_directory> [-v] <version> <ENV_NAME>
+
+		Options:
+		    -n    The name of the bundle to be updated
+		    -c    The directory with the content that will be used to create the bundle's update
+		    -v    The version where the update will be creted, if no version is selected the update
+		          will be created at the latest version found
+		EOM
+	}
+
+	# variables required by getopts
+	local OPTIND
+	local opt
+	# user provided
+	local env_name
+	local bundle_name
+	local version
+	local content_dir
+	# locally generated
+	local manifest
+	local web_dir
+	local target_dir
+	local state_dir
+	local version_dir
+	local old_version
+	local old_version_dir
+
+	# If no parameters are received show help
+	if [ $# -eq 0 ]; then
+		ubfc_usage
+		return
+	fi
+	set -f  # turn off globbing
+	while getopts :n:c:v: opt; do
+		case "$opt" in
+			n)	bundle_name="$OPTARG" ;;
+			v)	version="$OPTARG" ;;
+			c)	content_dir="$OPTARG" ;;
+			*)	ubfc_usage
+				return ;;
+		esac
+	done
+	set +f  # turn globbing back on
+	env_name=${*:$OPTIND:1}
+	validate_path "$env_name"
+	validate_path "$content_dir"
+	validate_param "$bundle_name"
+
+	echo "Updating bundle $bundle_name using content from $content_dir..."
+
+	# default values
+	# shellcheck disable=SC2010
+	# SC2010: Don't use ls | grep. Use a glob.
+	# Exception: ls has sorting options that are tricky
+	# to get right with other commands.
+	version=${version:-$(ls "$env_name"/web-dir | grep -E '^[0-9]+$' | sort -rn | head -n1)}
+	old_version=$(sudo cat "$env_name"/web-dir/"$version"/Manifest.MoM | grep -m1 "$bundle_name" | cut -f 3)
+
+	# shortcut to useful directories
+	web_dir="$env_name"/web-dir
+	target_dir="$env_name"/testfs/target-dir
+	state_dir="$env_name"/testfs/state
+	version_dir="$web_dir"/"$version"
+	old_version_dir="$web_dir"/"$old_version"
+
+	# to create the bundle update we will start treating the update as
+	# if it was a new bundle, then we will just adjust the generated artifacts
+	# to look like an update, so we will create a "new bundle" based on the new content first
+	create_bundle_from_content -n "$bundle_name" -v "$version" -c "$content_dir" "$env_name"
```

Comment:
The approach on using another function and undoing things that aren't necessary is not ideal.
The ideal is to find the things in common and create a function to perform that. And things that are different should be performed by each individual functions.